pkg/minishift/clusterup/clusterup.go

```diff
@@ -0,0 +1,227 @@
+	"github.com/minishift/minishift/pkg/minishift/oc"
+	"github.com/minishift/minishift/pkg/util"
+	"github.com/minishift/minishift/pkg/version"
+	"os"
+	"strings"
+	"time"
+)
+
+const (
+	ip_key             = "ip"
+	routing_suffix_key = "routing-suffix"
+)
+
+type ClusterUpConfig struct {
+	OpenShiftVersion string
+	MachineName      string
+	Ip               string
+	Port             int
+	RoutingSuffix    string
+	HostPvDir        string
+	User             string
+	Project          string
+	KubeConfigPath   string
+	OcPath           string
+}
+
+// ClusterUp downloads and installs the oc binary in order to run 'cluster up'
+func ClusterUp(config *ClusterUpConfig, clusterUpParams map[string]string, runner util.Runner) error {
+	cmdArgs := []string{"cluster", "up", "--use-existing-config"}
+
+	for key, value := range clusterUpParams {
+		if !oc.SupportFlag(key, config.OcPath, runner) {
+			errors.New(fmt.Sprintf("Flag %s is not supported for oc version %s. Use 'openshift-version' flag to select a different version of OpenShift.", key, config.OpenShiftVersion))
+		}
+		cmdArgs = append(cmdArgs, "--"+key)
+		cmdArgs = append(cmdArgs, value)
+	}
+
+	exitCode := runner.Run(os.Stdout, os.Stderr, config.OcPath, cmdArgs...)
+	if exitCode != 0 {
+		errors.New("Error starting the cluster.")
+	}
+	return nil
+}
+
+// PostClusterUp runs the Minishift specific provisioning after 'cluster up' has run
+func PostClusterUp(clusterUpConfig *ClusterUpConfig, sshCommander provision.SSHCommander, addOnManager *manager.AddOnManager) error {
+	err := kubeconfig.CacheSystemAdminEntries(clusterUpConfig.KubeConfigPath, getConfigClusterName(clusterUpConfig.Ip, clusterUpConfig.Port))
+	if err != nil {
+		return err
+	}
+
+	ocRunner, err := oc.NewOcRunner(clusterUpConfig.OcPath, clusterUpConfig.KubeConfigPath)
+	if err != nil {
+		return err
+	}
+
+	err = ocRunner.AddSudoerRoleForUser(clusterUpConfig.User)
+	if err != nil {
+		return err
+	}
+
+	err = ocRunner.AddCliContext(clusterUpConfig.MachineName, clusterUpConfig.Ip, clusterUpConfig.User, clusterUpConfig.Project)
+	if err != nil {
+		return err
+	}
+
+	err = configurePersistentVolumes(clusterUpConfig.HostPvDir, addOnManager, sshCommander, ocRunner)
+	if err != nil {
+		return err
+	}
+
+	err = applyAddOns(addOnManager, clusterUpConfig.Ip, clusterUpConfig.RoutingSuffix, ocRunner, sshCommander)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+//EnsureHostDirectoriesExist Ensures the specified directories exist on the virtual machine and creates them if not
+func EnsureHostDirectoriesExist(host *host.Host, dirs []string) error {
+	cmd := fmt.Sprintf("sudo mkdir -p %s", strings.Join(dirs, " "))
+	_, err := host.RunSSHCommand(cmd)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// DetermineOpenShiftVersion returns the OpenShift/oc version to use. If the requested version is
+// < the base line version we use baseline oc to provision the requested OpenShift version. If the requested
```

Comment:
Comma missing after `< the base line version`. Took me few  seconds to process this line :smile:  
Also, I feel if we write the comments one comment per line, then it will increase readability.

```
DetermineOpenShiftVersion returns the OpenShift/oc version to use. 
If the requested version is < the base line version we use baseline oc to provision the requested OpenShift version. 
If the requested OpenShift version is >= the baseline, we align the oc version with the requested OpenShift version.
```